core/src/apps/zcash/unified_addresses.py

```diff
@@ -0,0 +1,118 @@
+        return "u"
+    if coin.coin_name == "Zcash Testnet":
+        return "utest"
+    raise ValueError
+
+
+def padding(hrp: str) -> bytes:
+    assert len(hrp) <= 16
+    return hrp.encode() + bytes(16 - len(hrp))
+
+
+def encode(receivers: Dict[Typecode, bytes], coin: CoinInfo) -> str:
+    # multiple transparent receivers forbidden
+    assert not (Typecode.P2PKH in receivers and Typecode.P2SH in receivers)
+
+    length = 16  # 16 bytes for padding
+    for typecode in receivers.keys():
+        length += 2  # typecode (1 byte) + length (1 byte)
+        length += receiver_length(typecode)
+
+    w = empty_bytearray(length)
+
+    # receivers in decreasing order
+    receivers_list = list(receivers.items())
+    receivers_list.sort(reverse=True)
+
+    for (typecode, raw_bytes) in receivers_list:
+        write_compact_size(w, typecode)
+        write_compact_size(w, receiver_length(typecode))
+        write_bytes_fixed(w, raw_bytes, receiver_length(typecode))
+
+    hrp = prefix(coin)
+    write_bytes_fixed(w, padding(hrp), 16)
+    f4jumble(memoryview(w))
+    converted = convertbits(w, 8, 5)
+    return bech32_encode(hrp, converted, Encoding.BECH32M)
+
+
+def decode(addr_str: str, coin: CoinInfo) -> Dict[int, bytes]:
+    (hrp, data, encoding) = bech32_decode(addr_str, max_bech_len=1000)
+    if (hrp, data, encoding) == (None, None, None):
+        raise DataError("Bech32m decoding failed.")
+    assert hrp is not None  # to satisfy typecheckers
+    assert data is not None  # to satisfy typecheckers
+    assert encoding is not None  # to satisfy typecheckers
+    if hrp != prefix(coin):
+        raise DataError("Unexpected address prefix.")
+    if encoding != Encoding.BECH32M:
+        raise DataError("Bech32m encoding required.")
+
+    decoded = bytearray(convertbits(data, 5, 8, False))
+    f4unjumble(memoryview(decoded))
+
+    # check trailing padding bytes
+    if decoded[-16:] != padding(hrp):
+        raise DataError("Invalid padding bytes")
+
+    r = BufferReader(decoded[:-16])
+
+    last_typecode = None
+    receivers = {}
+    while r.remaining_count() > 0:
+        typecode = read_compact_size(r)
+        if typecode in receivers:
+            raise DataError("Duplicated typecode")
+        if typecode > 0x02000000:
+            raise DataError("Invalid typecode")
+        if last_typecode is not None and typecode > last_typecode:
+            raise DataError("Invalid receivers order")
+        last_typecode = typecode
+
+        length = read_compact_size(r)
+        # if the typecode of the receiver is known, then verify receiver length
+        try:
+            expected_length = receiver_length(typecode)
+            if length != expected_length:
+                raise DataError("Unexpected receiver length")
+        except ValueError:
+            pass
+
+        if r.remaining_count() < length:
+            raise DataError("Invalid receiver length")
+
+        receivers[typecode] = r.read(length)
+
```

Comment:
> Consumers MUST reject Unified Addresses/Viewing Keys [...] that include both P2SH and P2PKH Transparent Addresses, or that contain only a Transparent Address.

Technically we should be checking this and ideally have a unit test for each of these cases.

Reply:
adding the check [c7b9b7d](https://github.com/trezor/trezor-firmware/pull/2398/commits/c7b9b7d3954062246288f47b6dac644da0ee82b2)

Reply:
adding the second check [e780855](https://github.com/trezor/trezor-firmware/pull/2398/commits/e78085529fcc47d53fcf7045947f24d02837dcfc)

Reply:
fixing typechecking [163f493](https://github.com/trezor/trezor-firmware/pull/2398/commits/163f493d46ecf31c7f10077a1a9b102145ad6cae)